optimizely/cmab/cmab_client.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2025 Optimizely
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import time
+import requests
+import math
+from typing import Dict, Any, Optional
+from optimizely import logger as _logging
+from optimizely.helpers.enums import Errors
+
+# CMAB_PREDICTION_ENDPOINT is the endpoint for CMAB predictions
+CMAB_PREDICTION_ENDPOINT = "https://prediction.cmab.optimizely.com/predict/%s"
+
+# Default constants for CMAB requests
+DEFAULT_MAX_RETRIES = 3
+DEFAULT_INITIAL_BACKOFF = 0.1  # in seconds (100 ms)
+DEFAULT_MAX_BACKOFF = 10  # in seconds
+DEFAULT_BACKOFF_MULTIPLIER = 2.0
+MAX_WAIT_TIME = 10.0
+
+
+class CmabRetryConfig:
+    """Configuration for retrying CMAB requests.
+
+    Contains parameters for maximum retries, backoff intervals, and multipliers.
+    """
+    def __init__(
+        self,
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        initial_backoff: float = DEFAULT_INITIAL_BACKOFF,
+        max_backoff: float = DEFAULT_MAX_BACKOFF,
+        backoff_multiplier: float = DEFAULT_BACKOFF_MULTIPLIER,
+    ):
+        self.max_retries = max_retries
+        self.initial_backoff = initial_backoff
+        self.max_backoff = max_backoff
+        self.backoff_multiplier = backoff_multiplier
+
+
+class DefaultCmabClient:
+    """Client for interacting with the CMAB service.
+
+    Provides methods to fetch decisions with optional retry logic.
+    """
+    def __init__(self, http_client: Optional[requests.Session] = None,
+                 retry_config: Optional[CmabRetryConfig] = None,
+                 logger: Optional[_logging.Logger] = None):
```

Comment:
I wonder if we add timeout here (from requests library). because js sdk and go-sdk somewhat handle cacellation.
The JavaScript implementation doesn't explicitly handle cancellation like Go does, but it does use Promises which can be cancelled by the caller. Python should probably at least support configurable timeouts? maybe by adding a timeout arg? So if a request takes too long can be properly cancelled and processed as error/exception.

timeout: float = MAX_WAIT_TIME

and then self.timeout = timeout in the assignment.

Reply:
then in fetch_decision function you can do this:
```suggestion
                 
    def fetch_decision(self, rule_id, user_id, attributes, cmab_uuid, timeout=None):
        # Use the timeout parameter or fall back to default
        timeout = timeout or MAX_WAIT_TIME
    
        # Pass timeout to requests
        response = self.http_client.post(url, data=json.dumps(request_body), 
                                        headers=headers, timeout=timeout)
```